File: webhook.py
"""LINE Messaging APIのwebhookエンドポイント

"名前 <名前>" 形式のメッセージを受信した場合、
LINEユーザーIDと名前の対応をローカルのsqlite3データベースに保存する。

Usage:
    python webhook.py
"""
import base64
import hashlib
import hmac
import os
import re
import sqlite3
from pathlib import Path

import requests
from dotenv import load_dotenv
from fastapi import FastAPI, Header, HTTPException, Request

load_dotenv()

LINE_CHANNEL_SECRET = os.getenv("LINE_CHANNEL_SECRET")
LINE_CHANNEL_ACCESS_TOKEN = os.getenv("LINE_CHANNEL_ACCESS_TOKEN")
LINE_REPLY_URL = "https://api.line.me/v2/bot/message/reply"
DB_PATH = Path(__file__).parent / "users.db"
NAME_PATTERN = re.compile(r"^名前\s+(.+)$")
NGROK_ACCESS_TOKEN = os.getenv("NGROK_ACCESS_TOKEN")

app = FastAPI()


def init_db() -> None:
    with sqlite3.connect(DB_PATH) as conn:
        conn.execute(
            """
            CREATE TABLE IF NOT EXISTS users (
                user_id TEXT PRIMARY KEY,
                name TEXT NOT NULL
            )
            """
        )


def save_user_name(user_id: str, name: str) -> None:
    with sqlite3.connect(DB_PATH) as conn:
        conn.execute(
            """
            INSERT INTO users (user_id, name) VALUES (?, ?)
            ON CONFLICT(user_id) DO UPDATE SET name = excluded.name
            """,
            (user_id, name),
        )


def reply_message(reply_token: str, message: str) -> None:
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {LINE_CHANNEL_ACCESS_TOKEN}",
    }
    payload = {
        "replyToken": reply_token,
        "messages": [{"type": "text", "text": message}],
    }
    requests.post(LINE_REPLY_URL, headers=headers, json=payload)


def verify_signature(body: bytes, signature: str) -> bool:
    digest = hmac.new(LINE_CHANNEL_SECRET.encode(), body, hashlib.sha256).digest()
    expected_signature = base64.b64encode(digest).decode()
    return hmac.compare_digest(expected_signature, signature)


@app.on_event("startup")
def on_startup() -> None:
    init_db()


@app.post("/webhook")
async def webhook(request: Request, x_line_signature: str = Header(None)):
    body = await request.body()

    print("Signature",x_line_signature,verify_signature(body,x_line_signature))
    #if not x_line_signature or not verify_signature(body, x_line_signature):
    #    raise HTTPException(status_code=400, detail="Invalid signature")

    payload = await request.json()
    for event in payload.get("events", []):
        if event.get("type") != "message":
            continue

        message = event.get("message", {})
        if message.get("type") != "text":
            continue

        match = NAME_PATTERN.match(message.get("text", ""))
        user_id = event.get("source", {}).get("userId")
        if match and user_id:
            name = match.group(1).strip()
            save_user_name(user_id, name)

            reply_token = event.get("replyToken")
            if reply_token:
                reply_message(
                    reply_token,
                    f"名前のご登録ありがとうございます！「{name}」で登録されました。",
                )

    return {"status": "ok"}


if __name__ == "__main__":
    import uvicorn
    from pyngrok import ngrok

    port = int(os.getenv("PORT", "8000"))
    ngrok.set_auth_token(NGROK_ACCESS_TOKEN)
    public_url = ngrok.connect(port, "http")
    print(f"webhook URL: {public_url}/webhook")
    uvicorn.run(app, host="0.0.0.0", port=port)

File: webhook.js
/**
 * LINE Messaging APIのwebhookエンドポイント
 *
 * "名前 <名前>" 形式のメッセージを受信した場合、
 * LINEユーザーIDと名前の対応をローカルのsqlite3データベースに保存する。
 *
 * Usage:
 *     node webhook.js
 */
"use strict";

const crypto = require("crypto");
const path = require("path");
const { DatabaseSync } = require("node:sqlite");
const dotenv = require("dotenv");
const express = require("express");
const ngrok = require("@ngrok/ngrok");

dotenv.config({ quiet: true });

const LINE_CHANNEL_SECRET = process.env.LINE_CHANNEL_SECRET;
const DB_PATH = path.join(__dirname, "users.db");
const NAME_PATTERN = /^名前\s+(.+)$/;
const NGROK_ACCESS_TOKEN = process.env.NGROK_ACCESS_TOKEN;

const app = express();
app.use(express.raw({ type: "*/*" }));

function initDb() {
  const conn = new DatabaseSync(DB_PATH);
  conn.exec(`
    CREATE TABLE IF NOT EXISTS users (
        user_id TEXT PRIMARY KEY,
        name TEXT NOT NULL
    )
  `);
  conn.close();
}

function saveUserName(userId, name) {
  const conn = new DatabaseSync(DB_PATH);
  conn
    .prepare(
      `INSERT INTO users (user_id, name) VALUES (?, ?)
       ON CONFLICT(user_id) DO UPDATE SET name = excluded.name`
    )
    .run(userId, name);
  conn.close();
}

function verifySignature(body, signature) {
  const digest = crypto
    .createHmac("sha256", LINE_CHANNEL_SECRET)
    .update(body)
    .digest();
  const expectedSignature = digest.toString("base64");

  const expectedBuffer = Buffer.from(expectedSignature);
  const signatureBuffer = Buffer.from(String(signature));
  if (expectedBuffer.length !== signatureBuffer.length) return false;
  return crypto.timingSafeEqual(expectedBuffer, signatureBuffer);
}

app.post("/webhook", (req, res) => {
  const body = req.body;
  const xLineSignature = req.header("x-line-signature");

  console.log("Signature", xLineSignature, verifySignature(body, xLineSignature));
  // if (!xLineSignature || !verifySignature(body, xLineSignature)) {
  //   return res.status(400).json({ detail: "Invalid signature" });
  // }

  const payload = JSON.parse(body.toString("utf-8"));
  for (const event of payload.events || []) {
    if (event.type !== "message") continue;

    const message = event.message || {};
    if (message.type !== "text") continue;

    const match = NAME_PATTERN.exec(message.text || "");
    const userId = event.source && event.source.userId;
    if (match && userId) {
      saveUserName(userId, match[1].trim());
    }
  }

  res.json({ status: "ok" });
});

async function main() {
  initDb();

  const port = parseInt(process.env.PORT || "8000", 10);
  const listener = await ngrok.connect({ addr: port, authtoken: NGROK_ACCESS_TOKEN });
  console.log(`webhook URL: ${listener.url()}/webhook`);
  app.listen(port, "0.0.0.0");
}

if (require.main === module) {
  main();
}
